src/qibolab/instruments/icarusqfpga.py

```diff
@@ -1,259 +1,453 @@
+            dac_end_addr[dac] = max(end >> 4, dac_end_addr[dac])
+
+        payload = [
+            (dac, wfm, dac_end_addr[dac])
+            for dac, wfm in waveform_array.items()
+            if dac_end_addr[dac] != 0
+        ]
+        self.device.upload_waveform(payload)
+
+    def play_sequences(
+        self,
+        qubits: Dict[QubitId, Qubit],
+        couplers,
+        sequences: List[PulseSequence],
+        options: ExecutionParameters,
+    ):
+        pass
 
-            if retval:
-                return s.recv(1024).decode("utf-8")
+    def connect(self):
+        """Currently we only connect to the board when we have to send a
+        command."""
+        # Request the version from the board
+        ver = self.device.get_server_version()
+        log.info(f"Connected to {self.name}, version: {ver}")
 
-    def play_sequence(self):
-        self.fire()
+    def start(self):
+        pass
 
-    def connect(self):
+    def stop(self):
         pass
 
     def disconnect(self):
         pass
 
-    @staticmethod
-    def _hexify(pins):
-        return int(
-            "".join(["1" if i in set(pins) else "0" for i in reversed(range(24))]), 2
-        )
-
-
-class IcarusQFPGA(Instrument):
-    """Driver for the IcarusQ RFSoC socket-based implementation."""
-
-    def __init__(self, name, address, port=8080):
-        super().__init__(name, address)
-        self._dac_sample_size = 65536
-        self._adc_sample_size = 65536
-        self._dac_nchannels = 16
-        self._adc_nchannels = 16
-        self._dac_sampling_rate = 5898240000
-        self._adc_sampling_rate = 1966080000
-        self._nshots = 0
-
-        self.port = port
-        self._thread = None
-        self._buffer = None
-        self._adcs_to_read = None
-        self.nshots = None
-
-    def setup(self, dac_sampling_rate, adcs_to_read, **kwargs):
-        """Sets the sampling rate of the RFSoC. May need to be repeated several
-        times due to multi-tile sync error.
-
-        Arguments:
-            dac_sampling_rate_id (int): Sampling rate ID to be set on the RFSoC.
-            dac_sampling_rate_6g_id (int): Optional sampling rate ID for the 6GS/s mode if used.
-        """
-        self._adcs_to_read = adcs_to_read
-        self._dac_sampling_rate = dac_sampling_rate
-
-    def translate(self, sequence: List[Pulse], nshots):
-        """Translates the pulse sequence into a numpy array."""
+    def sweep(self):
+        pass
 
-        # Waveform is 14-bit resolution on the DACs, so we first create 16-bit arrays to store the data.
-        waveform = np.zeros((self._dac_nchannels, self._dac_sample_size), dtype="i2")
 
-        # The global time can first be set as float to handle rounding errors.
-        time_array = (
-            1 / self._dac_sampling_rate * np.arange(0, self._dac_sample_size, 1)
+class RFSOC_RO(RFSOC):
+    """IcarusQ RFSoC attached with readout capability."""
```

Comment:
Why `RFSOC` is handling ADCs and `READOUT` pulses as well, if in general has not readout capabilities?

Reply:
This is due to the board having both capabilities, but because we use much more DACs than ADCs, in a multiboard scenario, we will not use ADCs on one board.

I will eventually migrate to a cluster approach similar to what was done for Qblox.

Reply:
Sorry, my comment has been confusing.

I didn't want to sponsor the Qblox approach, not from the API point of view, and not even for the hardware. I'm not even against, simply my message was not about that.
Instead, I like the API to be as transparent as possible, and reflect the hardware, since Qibolab is already another layer on top.

Instead, my point was not much about why `RFSOC_RO` is doing both (which is fine), but why ADC is handled half in `RFSOC_RO` and half in the `RFSOC` base class.
I was highlighting that the distinction between the two is in the `_RO` version having readout capabilities, but even the base class (which then should not have ro capabilities in general) is handling part of the readout.